subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/dependencysubstitution/DefaultDependencySubstitutions.java

```diff
@@ -234,32 +234,41 @@ public Substitution withoutArtifactSelectors() {
 
             @Override
             public Substitution using(ComponentSelector notation) {
-                with(notation);
-                return this;
-            }
-
-            @Override
-            public void with(ComponentSelector substitute) {
-                DefaultDependencySubstitution.validateTarget(substitute);
+                DefaultDependencySubstitution.validateTarget(notation);
 
                 boolean projectInvolved = false;
-                if (substituted instanceof ProjectComponentSelector || substitute instanceof ProjectComponentSelector) {
+                if (substituted instanceof ProjectComponentSelector || notation instanceof ProjectComponentSelector) {
                     // A project is involved, need to be aware of it
                     projectInvolved = true;
                 }
 
                 if (substituted instanceof UnversionedModuleComponentSelector) {
                     final ModuleIdentifier moduleId = ((UnversionedModuleComponentSelector) substituted).getModuleIdentifier();
-                    if (substitute instanceof ModuleComponentSelector) {
-                        if (((ModuleComponentSelector) substitute).getModuleIdentifier().equals(moduleId)) {
+                    if (notation instanceof ModuleComponentSelector) {
+                        if (((ModuleComponentSelector) notation).getModuleIdentifier().equals(moduleId)) {
                             // This substitution is effectively a force
                             substitutionReason = substitutionReason.markAsEquivalentToForce();
                         }
                     }
-                    addSubstitution(new ModuleMatchDependencySubstitutionAction(substitutionReason, moduleId, substitute, () -> artifactAction), projectInvolved);
+                    addSubstitution(new ModuleMatchDependencySubstitutionAction(substitutionReason, moduleId, notation, () -> artifactAction), projectInvolved);
                 } else {
-                    addSubstitution(new ExactMatchDependencySubstitutionAction(substitutionReason, substituted, substitute, () -> artifactAction), projectInvolved);
+                    addSubstitution(new ExactMatchDependencySubstitutionAction(substitutionReason, substituted, notation, () -> artifactAction), projectInvolved);
                 }
+                return this;
+            }
+
+            @Override
+            @Deprecated
+            public void with(ComponentSelector substitute) {
+                // Do not nag for 7.1 as this introduces a performance regression with Android plugin
+                // Revisit when upgrading Android plugin
+                /*DeprecationLogger.deprecateMethod(Substitution.class, "with(ComponentSelector)")
+                    .replaceWith("using(ComponentSelector)")
+                    .willBeRemovedInGradle8()
+                    .withUpgradeGuideSection(7, "dependency_substitutions_with")
+                    .nagUser();*/
```

Comment:
Because emitting deprecation messages here currently cause performance regressions in Android builds, for now do not nag.
We should leave the issue open for now and revisit whether we can add the nagging here when upgrading Android plugin.

Reply:
Please create a ticket for this and add it to the comments.

Reply:
Left a note here https://github.com/gradle/gradle/issues/14107 - let's keep it open.

Reply:
sgtm